compliance-monitor/monitor.py

```diff
@@ -189,29 +212,82 @@ def import_bootstrap(bootstrap_path, conn):
         conn.commit()
 
 
-def precompute_targets(version):
-    suite = compile_suite(f"{version['version']}", version['include'])
-    return {
-        'suite': suite,
-        'targets': {
-            tname: suite.select(tname, target_spec)
+class PrecomputedVersion:
+    def __init__(self, version):
+        self.name = version['version']
+        self.suite = compile_suite(self.name, version['include'])
+        self.validity = version['validity']
+        self.targets = {
+            tname: self.suite.select(tname, target_spec)
             for tname, target_spec in version['targets'].items()
         }
-    }
+
+    def evaluate(self, scenario_results):
+        target_results = {}
+        for tname, suite in self.targets.items():
+            by_value = suite.evaluate(scenario_results)
+            target_results[tname] = {
+                'testcases': [testcase['id'] for testcase in suite.testcases],
+                'num_passed': len(by_value[1]),
+                'num_missing': len(by_value[0]),
+                'num_failed': len(by_value[-1]),
+                'result': -1 if by_value[-1] else 0 if by_value[0] else 1,
+            }
+        return {
+            'testcases': {tc['id']: tc for tc in self.suite.testcases},
+            'results': scenario_results,
+            'result': target_results['main']['result'],
+            'targets': target_results,
+            'validity': self.validity,
+        }
+
+
+class PrecomputedScope:
```

Comment:
What is a precomputed scope (and a precomputed version)? Should have small doc comment why this is needed. My understanding so far is, it's some kind of pre-warmed cache?

Reply:
I will add a comment. It could be construed as a pre-warmed cache. From my POV though, it's just the data structure that we need for our subsequent computations, and it's precomputed because it doesn't change in between those computations. I wouldn't necessarily call it a cache because (in general) every value is needed.

Reply:
4d6d491c2f4c94877a9bbb540b02aa93caeea137